src/webgpu/api/validation/copyTextureToTexture.spec.ts

```diff
@@ -210,3 +236,248 @@ g.test('multisampled_copy_restrictions')
+          { x: 0, y: 1, width: 0, height: 0 },
+          { x: 0, y: 0, width: -1, height: 0 },
+          { x: 0, y: 0, width: 0, height: -1 },
+        ])
+      )
+      .combine(
+        poptions('srcTextureSize', [
+          { width: 64, height: 64 },
+          { width: 64, height: 32 },
+          { width: 32, height: 32 },
+        ])
+      )
+      .combine(
+        poptions('dstTextureSize', [
+          { width: 64, height: 64 },
+          { width: 64, height: 32 },
+          { width: 32, height: 32 },
+        ])
+      )
+      .combine(poptions('srcCopyLevel', [1, 2]))
+      .combine(poptions('dstCopyLevel', [0, 1]))
+  )
+  .fn(async t => {
+    const {
+      format,
+      copyRectangleOffsets,
+      srcTextureSize,
+      dstTextureSize,
+      srcCopyLevel,
+      dstCopyLevel,
+    } = t.params;
+
+    const kMipLevelCount = 3;
+    const srcTexture = t.device.createTexture({
+      size: { width: srcTextureSize.width, height: srcTextureSize.height, depth: 1 },
+      format,
+      mipLevelCount: kMipLevelCount,
+      usage: GPUTextureUsage.COPY_SRC,
+    });
+    const dstTexture = t.device.createTexture({
+      size: { width: dstTextureSize.width, height: dstTextureSize.height, depth: 1 },
+      format,
+      mipLevelCount: kMipLevelCount,
+      usage: GPUTextureUsage.COPY_DST,
+    });
+
+    const srcTextureWidthAtLevel = srcTextureSize.width >> srcCopyLevel;
+    const srcTextureHeightAtLevel = srcTextureSize.height >> srcCopyLevel;
+    const dstTextureWidthAtLevel = dstTextureSize.width >> dstCopyLevel;
+    const dstTextureHeightAtLevel = dstTextureSize.height >> dstCopyLevel;
+
+    const copyOrigin = { x: copyRectangleOffsets.x, y: copyRectangleOffsets.y, z: 0 };
+
+    const copyWidth =
+      Math.min(srcTextureWidthAtLevel, dstTextureWidthAtLevel) +
+      copyRectangleOffsets.width -
+      copyOrigin.x;
+    const copyHeight =
+      Math.min(srcTextureHeightAtLevel, dstTextureHeightAtLevel) +
+      copyRectangleOffsets.height -
+      copyOrigin.y;
+
+    const isSuccess =
+      copyWidth === srcTextureWidthAtLevel &&
+      copyHeight === srcTextureHeightAtLevel &&
+      copyWidth === dstTextureWidthAtLevel &&
+      copyHeight === dstTextureHeightAtLevel;
+    t.TestCopyTextureToTexture(
+      { texture: srcTexture, origin: { x: 0, y: 0, z: 0 }, mipLevel: srcCopyLevel },
+      { texture: dstTexture, origin: copyOrigin, mipLevel: dstCopyLevel },
+      { width: copyWidth, height: copyHeight, depth: 1 },
+      isSuccess
+    );
+    t.TestCopyTextureToTexture(
+      { texture: srcTexture, origin: copyOrigin, mipLevel: srcCopyLevel },
+      { texture: dstTexture, origin: { x: 0, y: 0, z: 0 }, mipLevel: dstCopyLevel },
+      { width: copyWidth, height: copyHeight, depth: 1 },
+      isSuccess
+    );
+  });
+
+g.test('copy_ranges')
+  .params(
+    params()
+      .combine(
+        poptions('copyRectangleOffsets', [
+          { x: 0, y: 0, z: 0, width: 0, height: 0, depth: 1 },
```

Comment:
Aren't we missing testing of exactly corner cases? I'd assume for X we'd do the following:

```
          // full layer copy
          { x: 0, y: 0, z: 0, width: 0, height: 0, depth: 1 },
          // copy too big because x + width is larger than the width
          { x: 1, y: 0, z: 0, width: 0, height: 0, depth: 1 },
          { x: 0, y: 0, z: 0, width: 1, height: 0, depth: 1 },
//....
```

Reply:
Done